File: sem2/cpp/mergesort.cpp
#include <bits/stdc++.h>

using namespace std;

void merge(int a[],int l,int m,int r) {
	int L[m-l+1], R[r-m];
	for(int i=l; i<=m; i++)
		L[i-l] = a[i];
	for(int i=m+1; i<=r; i++)
		R[i-m-1] = a[i];
	int ln = m-l+1, rn = r-m, i=0, j=0, k=l;
	while(i<ln && j<rn) {
		if(L[i] > R[j]) {
			a[k] = R[j];
			j++;
		}
		else {
			a[k] = L[i];
			i++;
		}
		k++;
	}
	while(i < ln) {
		a[k] = L[i]; i++; k++;
	}
	while(j < rn) {
		a[k] = R[j]; j++; k++;
	}
}

void mergesort(int a[],int l,int r) {
	if(l >= r)
		return;
	int mid = (l+r)/2;
	mergesort(a, l, mid);
	mergesort(a, mid+1, r);
	merge(a, l, mid, r);
}

int main() {
	cout << "Enter the siize of the array:\n";
	int n;cin >> n;
	cout << "Enter the contents of the array:\n";
	int a[n];
	for(int i=0; i<n; i++) cin >> a[i];
	mergesort(a, 0, n-1);
	cout << "Sorted array:\n";
	for(int i=0; i<n; i++) cout << a[i] << " ";
	cout << endl;
	return 0;
}

File: sem2/python/mergesort.py
#!/usr/bin/python3

def merge(a, l, m, r):
    R,L = [],[]
    for i in a[m+1:r+1]:
        R.append(i)
    for i in a[l:m+1]:
        L.append(i)
    ln,rn = m-l+1,r-m
    i,j,k = 0,0,l
    while (i<ln and j < rn):
        if L[i] >= R[j]:
            a[k] = R[j]
            j+=1
        else:
            a[k] = L[i]
            i+=1
        k+=1
    while i < ln:
        a[k] = L[i]
        i+=1
        k+=1
    while j < rn:
        a[k] = R[j]
        j+=1
        k+=1

def merge_sort(a, l, r):
    if l >= r:
        return
    m = (l+r)//2
    merge_sort(a, l, m)
    merge_sort(a, m+1, r)
    merge(a, l, m, r)

a = list(map(int, input("Enter the contents of the array: ").split()))
merge_sort(a, 0, len(a)-1)
print("Sorted array: ", *a)
